blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
First, these are not just minor changes you had to make in `Softmax` to make it handle the time dimension. Its code becomes considerably more involved, and people not needing this feature will have harder to time to understand it. Still most people work with FF nets. 

Second, there are more potential usecases for this brick than just softmax. It can wrap any operation that can act on batches of tuples, a vector and an integer scalar in the softmax case, for which writing an implementation unaware of extra axis is just simpler. But for `tensor.dot` magical ability to handle extra-axis, `Linear` and `MLP` would be the first consumers for `WithExtraDims`. Many but not all Theano expressions are as agnostic to additional axis, e.g. [fft](https://github.com/Theano/Theano/blob/master/theano/sandbox/fourier.py#L96) does not seem to be the case.

I agree that it is tempting to wrap the whole brick instead of only application method. This does not seem to be possible codewise though. We can't add application methods to a brick at instantiation time, for one. Dispatching the application method to be called by its name would be quite ugly, since I don't like application methods that do very different things depending on what their parameters are. I guess this is one of those case when we have to tolerate that one brick with multiple application method will be references as a child by a set of bricks wrapping each of these methods.

It is true that `WithExtraDims` puts certain restrictions on the inputs and outputs, but these correspond to the major case when we just have a batch of data and want to act element-wise, but the indices in the batch are 2+ dimensional.